activerecord/lib/active_record/transactions.rb

```diff
@@ -224,6 +224,41 @@ def transaction(**options, &block)
         end
       end
 
+      class CallbackCountDownLatch # :nodoc:
+        def initialize(count, callback)
+          raise ArgumentError, "count must be at least 2" if count < 2
+          @count = count
+          @callback = callback
+        end
+
+        def call
+          @count -= 1
+          @callback.call if @count.zero?
+        end
+      end
+
+      def after_current_transaction_commit(&block)
+        open_transactions = []
+        connection_handler.each_connection_pool do |pool|
+          if active_connection = pool.active_connection
+            if active_connection.current_transaction.open? && active_connection.current_transaction.joinable?
+              open_transactions << active_connection.current_transaction
+            end
+          end
+        end
+
+        if open_transactions.empty?
+          yield
+        elsif open_transactions.size == 1
+          open_transactions.first.after_commit(&block)
+        else
+          callback = count_down_callback(open_transactions.size, &block)
```

Comment:
Possibly tying in with my comment above 😅 -- I'm not sure what I think _should_ happen, but I think this will behave surprisingly if any of the transactions get rolled back (and thus `count` never reaches zero).

Reply:
That's how I see it though. If any transaction is rolled back, we don't enqueue the job at all. I think that's what we want most of the time, no?

Reply:
I understand that we'll probably need an escape hatch, but e.g.:

```ruby
transaction do
  user = User.create!
  WelcomeEmailJob.perform_later(user)
  # something that fails and rollback the transaction
end
```

If the transaction is rolled back, you don't want this job enqueued.

Now maybe `perform_later` should be left unchanged, and it's another API that should have this behavior.

But the goal here is to reduce the behavioral difference between queues that use the same DB as Active Record, hence are transaction aware, and those that aren't.

In the above example, if you are using SolidQueue / GoodJob etc, the `perform_later` is acked but will be dropped. So I think we're consistent here.